[RFC] Partial reload support (#14640)
open-telemetry/opentelemetry-collector: docs/rfcs/partial-reload.md

Reviewer: I think it is worth mentioning how this relates to the component instancing rules we have, discussed e.g. on the motivation section of #10534. I feel like the rules described here are consistent with that model, but it seems worth describing this explicitly somewhere

# Partial Reload for the OpenTelemetry Collector

Support partial configuration reload to rebuild only affected components, reducing event loss and downtime during configuration changes. Rather than tearing down and rebuilding the entire pipeline graph on every configuration update, the collector will diff the old and new configurations and selectively restart only the components that have changed. This will be rolled out incrementally behind a feature flag, starting with receivers and expanding to all component types.

> ## Motivation

Reviewer 2: If we do this, I think it will likely require the `Reload` interface you mention below, so I'm okay to leave it out for now. However, in the future, I would like to cover that case in the situation where a client fires a request and quits, which will be particularly important for mobile/web clients. I agree though that most clients likely are running on user-controlled infrastructure and can do retries.

Currently, when a configuration change is applied to the OpenTelemetry Collector, the entire pipeline graph is torn down and rebuilt from scratch. This means every receiver, processor, exporter, and connector is stopped and restarted, regardless of whether it was affected by the change. This results in costly re-initialization of components, loss of accumulated in-memory state, disruption of consumption cursors and checkpoints, and unnecessary impact on external systems.

> A survey of components in the collector-contrib repository reveals the full scope of this problem across four categories of impact:

Reviewer: Seems worth mentioning that a number of these components are in our priority list (https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/44130) because we think they are among the most used components

Author: I mentioned that many are in the priority list.

### Expensive Startup

> Components with costly initialization that involves large state synchronization, service discovery, or log replay. Restarting these components introduces significant delays before they are fully operational.

Reviewer 2: >significant delays

I know there are a number of factors that influence this, but do we have any way to quantify what's considered significant here? If it takes e.g. <1 second to start these, I think we would need to justify why that's considered a significant amount of time.

Reviewer 2: Thanks for the info. I think this may be worth putting in the RFC, but if there is broad agreement this is valuable even without these metrics, I'm okay leaving it out.

Author: I added this information to make it more clear.

- `k8sclusterreceiver` - performs a full list/watch sync of the Kubernetes cluster state on startup, with a default timeout of 10 minutes. No metrics are emitted until the sync completes.
- `prometheusreceiver` - re-runs all service discovery backends (Kubernetes SD, Consul SD, DNS SD, etc.) and rebuilds per-target scrape loops from scratch.
- `prometheusremotewriteexporter` - replays unacknowledged write-ahead log (WAL) entries on startup before accepting new data, adding latency proportional to the WAL backlog.
- `k8sattributesprocessor` - when configured with `wait_for_metadata: true`, blocks startup until its Kubernetes metadata cache is fully synced via informers, delaying the pipeline from accepting data.

### Accumulated State Loss

> Many components maintain in-memory state (buffers, counters, accumulators, caches) that is permanently lost on restart. This causes metric gaps, trace loss, incorrect calculations, and false alerts visible to end users.

Reviewer 2: Could you mention this as a motivating factor? I think it's worth calling out given that a persistence layer is arguably the more robust solution, but comes with the tradeoff of needing the necessary infrastructure to persist data.

> Many components maintain in-memory state (buffers, counters, accumulators, caches) that is permanently lost on restart. This causes metric gaps, trace loss, incorrect calculations, and false alerts visible to end users.

Author: I added this to the RFC.

- `statsdreceiver` - accumulates gauge, counter, histogram, and timing values from incoming StatsD packets between flush intervals. On restart, any partial aggregations not yet flushed are discarded.
- `prometheusexporter` - maintains a metric registry storing the last known value for all time series exposed on the scrape endpoint. On restart, the endpoint returns empty metrics until new data arrives, potentially breaking alerting and dashboard rules.
- `signalfxexporter` - buffers pending dimension metadata updates and correlation state for deduplication. On restart, pending dimension updates are lost and metadata associations must be re-established.
- `k8sattributesprocessor` - maintains an in-memory metadata cache of Kubernetes resources (pods, namespaces, nodes, deployments) populated via informers. By default, the cache is refilled asynchronously so startup is not blocked, but data flowing through the processor during this period will have its Kubernetes metadata missing. With `wait_for_metadata: true`, the processor delays startup until the cache is fully synced, trading availability for correctness.

### Cursor and Checkpoint Disruption

> Components that track consumption progress via file offsets, poll cursors, or partition checkpoints. When configured with a storage extension, these components persist their position and can recover on restart. Without a storage extension, positions are held in memory and lost on restart, causing duplicate ingestion or missed data.

Reviewer 2: I think the second paragraph is particularly persuasive, could you add that to this section?

> Components that track consumption progress via file offsets, poll cursors, or partition checkpoints. When configured with a storage extension, these components persist their position and can recover on restart. Without a storage extension, positions are held in memory and lost on restart, causing duplicate ingestion or missed data.

Author: Added more of this to the RFC.

- `awscloudwatchreceiver` - stores timestamp checkpoints per log group. On restart without storage, checkpoints reset to the configured `start_from` time, causing duplicate ingestion of logs already consumed.
- `mongodbatlasreceiver` - stores the last recorded alert timestamp for polling. On restart without storage, the timestamp is lost and the receiver re-fetches alerts from the full polling window, causing alert duplication.

### External System Disruption

Components whose restart causes disruption to external systems or other instances beyond the collector itself.

- `kafkareceiver` - shutting down the receiver sends a LeaveGroup request to the Kafka broker, triggering a consumer group rebalance that affects all members of the group.

---

By enabling partial reload, the collector can intelligently determine which components are affected by a configuration change and only restart those specific components. Pipelines and components that are unaffected by the change can continue operating without interruption, significantly reducing the impact of configuration updates.

## Explanation

> When partial reload is enabled via the `service.partialReload` feature flag, the collector compares the incoming configuration with the current running configuration to determine the minimal set of components that need to be rebuilt. Components not affected by the change continue running without interruption, unless they are upstream of a changed component in the same pipeline, in which case they are also rebuilt to ensure a clean data path.

Reviewer 2: >unless they are upstream of a changed component in the same pipeline, in which case they are also rebuilt to ensure a clean data path

I discussed this with @swiatekm and it sounds like it's just a byproduct of the way our pipelines work, but is this a strict technical limitation? Ideally we could have a way to manipulate the consumers in-place and drain any items in the old component instance.

Author: Correct. This is only described this way in the RFC because of how the pipelines work today. It was an attempt to not add locking and ensure that any events flowing through the pipeline don't get lost or miss communicated. Work could be done in the graph to remove the need to re-build upstream for a component change. We would just need to be caution of the implications of such a change. Currently doing a full restart of the affected upstream components is the simplest and less risk adverse change currently.

Reviewer 3: It's not a strict technical limitation. Components are connected through function calls. If we want to replace a component without also replacing components upstream of it, then the upstream components need a way to switch out the consumer function they're using. This would require introducing locking, which may have performance implications, but I don't think anyone has ever benchmarked this seriously.

That said, I imagine orchestrating this kind of switch without losing data would be quite complex. That's one of the reasons it's not in scope for this RFC.

Author: I added to the RFC an alternative implementation that will allow the replacement of components in the graph without having the re-create the upstream components when a reload occurs. See this gate PR for an implementation on how this could be used in the graph construction. https://github.com/open-telemetry/opentelemetry-collector/pull/15254

> ### Component Changes

Reviewer: I am going to be a bit pedantic on the next set of comments just because I think it is useful to be strict in distinguishing between 'component configuration' and 'component instance' on the text, bear with me :)

```suggestion
### Component Configuration Changes
```

> - **Receiver modified**: Only the affected receiver is rebuilt. Downstream processors and exporters continue operating and will receive data from the new receiver instance once it starts.

Author: Great feedback, thanks!

- **Processor modified**: The processor and all processors and receivers upstream of it in the pipeline are rebuilt. This ensures that no in-flight data is lost during the processor restart.

> - **Exporter modified**: The exporter, along with all processors and receivers upstream of it in the pipeline, are rebuilt. This provides a clean data path from ingestion to export.

Reviewer: ```suggestion
- **Receiver configuration modified**: Only the affected receiver instance(s) are rebuilt. Downstream processors and exporter instances continue operating and will receive data from the new receiver instance once it starts.
- **Processor configuration modified**: The processor instance(s) and all processors and receiver instances upstream of it in any pipelines are rebuilt. This ensures that no in-flight data is lost during the processor instance(s) restart.
- **Exporter configuration modified**: The exporter instance(s), along with all processors and receiver instances upstream of it in any pipelines, are rebuilt. This provides a clean data path from ingestion to export.
```

> ### Pipeline Changes

Reviewer: ```suggestion
### Pipeline Configuration Changes
```

> - **Pipeline added**: Only the new pipeline is instantiated and started. Existing pipelines continue operating without interruption.

Reviewer: I guess it is worth mentioning that this may include adding existing component instances to the new pipeline.

Author: That is actually discussed in the RFC for in each type. So adding a receiver is discussed in the receivers section, etc for each section.

- **Pipeline removed**: Only the removed pipeline is stopped and torn down. Other pipelines remain unaffected.
- **Pipeline modified**: When a receiver, processor, or exporter is added to or removed from an existing pipeline, only that pipeline is rebuilt. Other pipelines remain unaffected.

### Connector Changes

Connectors require special handling since they bridge two pipelines:

- **Connector added**: On the exporter side (the pipeline feeding into the connector), the entire upstream pipeline is rebuilt to establish the new data flow. On the receiver side (the pipeline receiving from the connector), the connector is started as a new receiver and downstream processors and exporters continue operating without interruption.

> - **Connector removed**: On the exporter side, the pipeline is rebuilt to remove the connector as a destination. On the receiver side, the connector is stopped. If the receiving pipeline has other receivers, they continue operating unaffected.

Reviewer: ```suggestion
- **Connector added**: On the exporter side (the pipeline feeding into the connector), the entire upstream pipeline is rebuilt to establish the new data flow. On the receiver side (the pipeline receiving from the connector), the connector instance is started as a new receiver and downstream processors and exporter instances continue operating without interruption.
- **Connector removed**: On the exporter side, the pipeline is rebuilt to remove the connector instance as a destination. On the receiver side, the connector instance is stopped. If the receiving pipeline has other receivers, they continue operating unaffected.
```

## Internal details

### Feature Flag

The partial reload functionality will be entirely behind a feature flag named `service.partialReload`. Only when this flag is explicitly enabled will the new partial reload code path be taken. When disabled, the collector continues to use the existing full reload behavior. This ensures that no existing behavior is changed by default.

### Configuration Diff

On each configuration change event, the collector performs a diff between the current and new configurations using `reflect.DeepEqual` to keep the comparison simple:

1. Compare service-level settings (telemetry, extensions). If any have changed, fall back to full reload.
2. Compare each component type (receivers, processors, exporters, connectors) to identify which specific component instances have been added, removed, or modified.
3. For each pipeline, determine if the pipeline itself has changed (added, removed, or its component references modified).
4. Based on the diff, calculate the minimal set of components to rebuild using the "up the graph" propagation rule.

### Phased Rollout

The implementation will follow a phased approach to minimize risk:

> Each phase will be validated independently before moving to the next. When a phase is not yet implemented, changes to those component types will fallback to a full reload.

Reviewer: Howe will the validation be done? Do we expect that we would have separate feature gates for each phase (it seems like we only want one, but that would be all-or-nothing)? Do we expect any downstream distros to adopt this and give us feedback?

Author: I think it was discussed of using beta feature flags that are enabled by default for each stage of the partial reload: receivers, processors, exporters, pipelines, etc. Then use an alpha gate for enabling all of partial reload, which is off by default. Meaning that if you turn on partial reload, you get all the beta flags on by default, but can turn them off specifically if you have partial reload enabled.

#### Phase 5 - Connectors

Partial reload support for connectors, including adding, removing, and modifying their configurations. Connectors bind two pipelines together, acting as an exporter in one pipeline and a receiver in another. Because of this, changes to a connector affect both pipelines. On the exporter side, the entire upstream pipeline is rebuilt to establish or remove the data flow. On the receiver side, the portion of the pipeline starting from the connector is rebuilt.

### Error Handling

- If partial reload fails mid-way (e.g., a new receiver fails to start), the graph may be in a partially modified state. In this case, the error is propagated and the collector exits, consistent with how full reload failures are handled today.
- If the configuration diff determines that the change cannot be handled by the currently implemented phase, the collector falls back to a full reload transparently.

### Corner Cases

> - **Shared components**: A receiver instance may be shared across multiple pipelines. The diff logic must account for this and only rebuild the receiver once, reconnecting it to all relevant pipelines.

Reviewer: How will the diff logic account for this? Will we label which components share instance across multiple pipelines? Would be good to also have a list of what components do this today (we must deal with this because the OTLP receiver handles this, but it would be good to know the extent of the problem)

Reviewer 3: I don't think there's a problem there beyond the existing issues with shared component status reporting. The only new effect there would be one of the sharing components being shut down and then started without the rest being touched. With full reload, this always happens in waves - we shut down everything, presumably drop the shared bits, and then recreate it. The current shared component implementations already have to deal with these semantics because starting the graph is asynchronous, but partial reload might surface hitherto unknown concurrency bugs.

## Trade-offs and mitigations

**Increased code complexity**: The partial reload logic adds new code paths for diffing configurations and selectively rebuilding components. This is mitigated by:
- Gating behind a feature flag so the new paths are only exercised when opted in.
- Not modifying any existing core reload logic. The partial reload is an entirely separate code path.
- Phased rollout that allows each component type to be validated independently.

**Partial failure states**: If a component fails to start during partial reload, the pipeline graph may be in an inconsistent state. This is mitigated by treating partial reload failures the same as full reload failures (propagate the error). Future work could explore rollback semantics.

**Configuration comparison overhead**: Diffing configurations on every change adds overhead. In practice, this overhead is negligible compared to the cost of tearing down and rebuilding components, and uses `reflect.DeepEqual` on configuration structs which is fast for typical collector configurations.

## Open questions

- What metrics and observability should be added to track partial reload performance and success rates?

> - Should partial reload eventually become the default behavior, replacing full reload entirely, or should it remain opt-in behind a feature flag?

Reviewer 2: ```suggestion
- Should partial reload eventually become the default behavior, replacing full reload entirely, or should it remain opt-in behind a config flag?
```
I think long-term we'll want a config flag instead of a feature gate, since our feature gates are typically meant to have a limited lifespan. Open to other arguments here.

Author: That means we keep partial reload as always an option enablement or does it just become the default operating mode in the future?

Author: Feature flag will be used for now, we can look at switching to a config option in the future if others are interested in that way of enablement.

## Future possibilities

> - **Component hot reload**: Building on partial reload, a future enhancement could allow components to implement a `Reload` interface that accepts new configuration without requiring a stop/start cycle, further reducing disruption.

Reviewer 4: It would be also great for db receivers, as some db receivers, https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/sqlserverreceiver, have internal cache to maintain the correct result.

Related ticket: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/47514

> - **Component hot reload**: Building on partial reload, a future enhancement could allow components to implement a `Reload` interface that accepts new configuration without requiring a stop/start cycle, further reducing disruption.

Reviewer 4: Raised a proposal for the hot reload. https://github.com/open-telemetry/opentelemetry-collector/issues/15228

> - **Component hot reload**: Building on partial reload, a future enhancement could allow components to implement a `Reload` interface that accepts new configuration without requiring a stop/start cycle, further reducing disruption.

Author: Great to hear that it would be great to have. It would definitely be in the journey for full partial reload, but all this stuff needs to happen first before that really can be supported.

> - **Component hot reload**: Building on partial reload, a future enhancement could allow components to implement a `Reload` interface that accepts new configuration without requiring a stop/start cycle, further reducing disruption.

Reviewer 4: > but all this stuff needs to happen first before that really can be supported.

@blakerouse I think the hot reload can go parallelly with partial reload without being blocked. Hot reload can focus on replacing `stop/start` cycle with `reload`. If partial reload is not there, it would try to reload every component.

You can see my POC code (still need iteration) and it can be done without partial reload.

> - **Component hot reload**: Building on partial reload, a future enhancement could allow components to implement a `Reload` interface that accepts new configuration without requiring a stop/start cycle, further reducing disruption.

Reviewer 4: @blakerouse I found reloading components other than receivers are difficult. I think you are right. The hot reload is relied on partial reload.

> - **Component hot reload**: Building on partial reload, a future enhancement could allow components to implement a `Reload` interface that accepts new configuration without requiring a stop/start cycle, further reducing disruption.

Author: @XSAM Yeah I believe we need this partial reload first, then we can add the `Reload` interface to allow components that want to support it.

## Motivation

Currently, when a configuration change is applied to the OpenTelemetry Collector, the entire pipeline graph is torn down and rebuilt from scratch. This means every receiver, processor, exporter, and connector is stopped and restarted, regardless of whether it was affected by the change. This results in costly re-initialization of components, loss of accumulated in-memory state, disruption of consumption cursors and checkpoints, and unnecessary impact on external systems.

A survey of components in the collector-contrib repository reveals the full scope of this problem across four categories of impact (many are in the priority for stabilization as well: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/44130):

### Expensive Startup

Components with costly initialization that involves large state synchronization, service discovery, or log replay. Restarting these components introduces significant delays before they are fully operational.

> - `k8sclusterreceiver` - performs a full list/watch sync of the Kubernetes cluster state on startup, with a default timeout of 10 minutes. No metrics are emitted until the sync completes. Seems it can take 2-5 minutes, so 10 minutes was picked to be long enough (https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/842#discussion_r488211782).

Reviewer 5: Let's use config name instead of go package name, e.g. `k8s_cluster` receiver. Maybe that will fix `spell-check`

Author: I have updated it to use the config names.

### Alternative To Restarting Upstream Components

The above section highlights an implementation detail of restarting upstream components when a downstream component needs to be replaced. This is mentioned as a way of implementation as it allows no changes to be made to the current graph structure. An alternative approach is adding a gate between each component in the pipeline. The gate would allow pausing the flow of events (allowing those that already passed through the gate to return) which facilitates the ability to replace the components in the pipeline downstream of the gate without re-constructing the upstream components. See: https://github.com/open-telemetry/opentelemetry-collector/pull/15254 for gate design and benchmarks.

### Telemetry and Extension Changes

Changes to telemetry or extension configurations will always result in a full reload, as these are service-level settings that affect the entire collector.

## Internal details

### Feature Flag

> The partial reload functionality will be entirely behind a feature flag named `service.partialReload`. Only when this flag is explicitly enabled will the new partial reload code path be taken. When disabled, the collector continues to use the existing full reload behavior. This ensures that no existing behavior is changed by default.

Reviewer: ```suggestion
The partial reload functionality will be entirely behind a feature flag named `service.partialReload`. Only when this flag is explicitly enabled will the new partial reload code path be taken. When disabled, the collector continues to use the existing full reload behavior. This ensures that no existing behavior is changed by default.

We may use multiple feature gates for gradually enabling functionality as it becomes battle tested
```
